### ISSUE TITLE
Release assembly with weaved Task function crashes 

### ISSUE THREAD

**Issue by hhblaze** [ArxOne/MrAdvice#106]
Hi, another interesting bird.
I add the complete, as simple as possible, project emulating bug.
When you start it from VS in Debug mode - works, when you start exe from DEBUG folder - works.
When you start it from VS in Release mode or exe from from RELEASE folder - crashes. with Critical error (the same as in previous bug).

[TestAsync.zip](https://github.com/ArxOne/MrAdvice/files/1038099/TestAsync.zip)

Both not working neither async function calling async Task nor sync calling async
```C#
//Fatal error
 public void StartTask()
            {
                var task = System.Threading.Tasks.Task.Run(async () => await AddWhatever());
                var r = task.Result;                
            }

//Fatal error
            public async void StartTask1()
            {
                long r = await AddWhatever();
            }


 public async System.Threading.Tasks.Task<long> AddWhatever()
            {
                return -1;
            }

```

**Comment by picrap** [comment]
I need more time to dig the problem.
However, you can apply the attribute to your advices, in order to exclude the async state machine classes from being weaved.

In your sample, apply 
`[ArxOne.MrAdvice.Annotation.ExcludePointcut("*+<@>d@.@")]` 
to your `PerformanceAspect` class.

And it will still run in release.

**Comment by hhblaze** [comment]
Thanx, will work around for now

**Comment by picrap** [comment]
Can you test version 2.5.9 and confirm it works without workaround?

**Comment by hhblaze** [comment]
Thank you, will do, and write if I have troubles with that

**Comment by hhblaze** [comment]
Ok, test project is compiling and running with 2.5.11 like it should be, but my original fails on weaving time with errors:

1>...path to...\packages\MrAdvice.2.5.11\build\MrAdvice.targets(10,5): error : Internal error during Weavable methods detection: System.Runtime.Serialization.SerializationException: Type 'dnlib.DotNet.MethodDefMD' in Assembly 'MrAdvice.Weaver, Version=2.0.0.0, Culture=neutral, PublicKeyToken=c0e7e6eab6f293d8' is not marked as serializable.
1>...path to...\packages\MrAdvice.2.5.11\build\MrAdvice.targets(10,5): error : 
1>...path to...\packages\MrAdvice.2.5.11\build\MrAdvice.targets(10,5): error : Server stack trace: 
1>...path to...\packages\MrAdvice.2.5.11\build\MrAdvice.targets(10,5): error :    at System.Runtime.Serialization.FormatterServices.InternalGetSerializableMembers(RuntimeType type)
1>...path to...\packages\MrAdvice.2.5.11\build\MrAdvice.targets(10,5): error :    at System.Runtime.Serialization.FormatterServices.<>c__DisplayClass9_0.<GetSerializableMembers>b__0(MemberHolder _)
1>...path to...\packages\MrAdvice.2.5.11\build\MrAdvice.targets(10,5): error :    at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
1>...path to...\packages\MrAdvice.2.5.11\build\MrAdvice.targets(10,5): error :    at System.Runtime.Serialization.FormatterServices.GetSerializableMembers(Type type, StreamingContext context)
1>...path to...\packages\MrAdvice.2.5.11\build\MrAdvice.targets(10,5): error :    at System.Runtime.Serialization.Formatters.Binary.WriteObjectInfo.InitMemberInfo()
1>...path to...\packages\MrAdvice.2.5.11\build\MrAdvice.targets(10,5): error :    at System.Runtime.Serialization.Formatters.Binary.WriteObjectInfo.InitSerialize(Object obj, ISurrogateSelector surrogateSelector, StreamingContext context, SerObjectInfoInit serObjectInfoInit, IFormatterConverter converter, ObjectWriter objectWriter, SerializationBinder binder)
1>...path to...\packages\MrAdvice.2.5.11\build\MrAdvice.targets(10,5): error :    at System.Runtime.Serialization.Formatters.Binary.WriteObjectInfo.Serialize(Object obj, ISurrogateSelector surrogateSelector, StreamingContext context, SerObjectInfoInit serObjectInfoInit, IFormatterConverter converter, ObjectWriter objectWriter, SerializationBinder binder)
1>...path to...\packages\MrAdvice.2.5.11\build\MrAdvice.targets(10,5): error :    at System.Runtime.Serialization.Formatters.Binary.ObjectWriter.Serialize(Object graph, Header[] inHeaders, __BinaryWriter serWriter, Boolean fCheck)
1>...path to...\packages\MrAdvice.2.5.11\build\MrAdvice.targets(10,5): error :    at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Serialize(Stream serializationStream, Object graph, Header[] headers, Boolean fCheck)
1>...path to...\packages\MrAdvice.2.5.11\build\MrAdvice.targets(10,5): error :    at System.Runtime.Remoting.Channels.CrossAppDomainSerializer.SerializeMessageParts(ArrayList argsToSerialize)
1>...path to...\packages\MrAdvice.2.5.11\build\MrAdvice.targets(10,5): error :    at System.Runtime.Remoting.Messaging.SmuggledMethodCallMessage..ctor(IMethodCallMessage mcm)
1>...path to...\packages\MrAdvice.2.5.11\build\MrAdvice.targets(10,5): error :    at System.Runtime.Remoting.Messaging.SmuggledMethodCallMessage.SmuggleIfPossible(IMessage msg)
1>...path to...\packages\MrAdvice.2.5.11\build\MrAdvice.targets(10,5): error :    at System.Runtime.Remoting.Channels.CrossAppDomainSink.SyncProcessMessage(IMessage reqMsg)
1>...path to...\packages\MrAdvice.2.5.11\build\MrAdvice.targets(10,5): error : 
1>...path to...\packages\MrAdvice.2.5.11\build\MrAdvice.targets(10,5): error : Exception rethrown at [0]: 
1>...path to...\packages\MrAdvice.2.5.11\build\MrAdvice.targets(10,5): error :    at System.Runtime.Remoting.Proxies.RealProxy.HandleReturnMessage(IMessage reqMsg, IMessage retMsg)
1>...path to...\packages\MrAdvice.2.5.11\build\MrAdvice.targets(10,5): error :    at System.Runtime.Remoting.Proxies.RealProxy.PrivateInvoke(MessageData& msgData, Int32 type)
1>...path to...\packages\MrAdvice.2.5.11\build\MrAdvice.targets(10,5): error :    at StitcherBoy.Logging.ILogging.WriteDebug(String format, Object[] parameters)
1>...path to...\packages\MrAdvice.2.5.11\build\MrAdvice.targets(10,5): error :    at ArxOne.MrAdvice.Weaver.AspectWeaver.IsFromComputerGeneratedType(MarkedNode markedMethod)
1>...path to...\packages\MrAdvice.2.5.11\build\MrAdvice.targets(10,5): error :    at ArxOne.MrAdvice.Weaver.AspectWeaver.<Weave>b__24_0(MarkedNode n)
1>...path to...\packages\MrAdvice.2.5.11\build\MrAdvice.targets(10,5): error :    at System.Linq.Enumerable.WhereEnumerableIterator`1.MoveNext()
1>...path to...\packages\MrAdvice.2.5.11\build\MrAdvice.targets(10,5): error :    at System.Linq.Buffer`1..ctor(IEnumerable`1 source)
1>...path to...\packages\MrAdvice.2.5.11\build\MrAdvice.targets(10,5): error :    at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
1>...path to...\packages\MrAdvice.2.5.11\build\MrAdvice.targets(10,5): error :    at ArxOne.MrAdvice.Weaver.AspectWeaver.Weave(ModuleDefMD moduleDefinition)
1>...path to...\packages\MrAdvice.2.5.11\build\MrAdvice.targets(10,5): error : Please complain, whine, cry, yell at https://github.com/ArxOne/MrAdvice/issues/new



If you need to open new issue for that let me know

**Comment by picrap** [comment]
I am stupid! Give me a few hours and I'll fix this. Sorry.

**Comment by hhblaze** [comment]
With 2.5.1 the same code perfectly weaves

**Comment by hhblaze** [comment]
ok, waiting so much!!!

**Comment by picrap** [comment]
With 2.5.1, have you kept the `[ExcludePointcut]` ? Because the fix is not present before 2.5.9...

**Comment by hhblaze** [comment]
With 2.5.1 it weaves with and without [ExcludePointcut].
Of course my current project with 2.5.1 has  [ExcludePointcut].

2.5.11 doesn't weave with and without  [ExcludePointcut].

**Comment by picrap** [comment]
You can try 2.5.12 once it's released (it's currently building).

**Comment by hhblaze** [comment]
Looks like working!
Thank you!

**Comment by hhblaze** [comment]
The other project could not be weaved (2.5.12):

2>...path to...\packages\MrAdvice.2.5.12\build\MrAdvice.targets(10,5): error : Unhandled exception: dnlib.DotNet.Writer.ModuleWriterException: Instruction has been removed but it's referenced by PDB info
2>...path to...\packages\MrAdvice.2.5.12\build\MrAdvice.targets(10,5): error :    at dnlib.DotNet.DummyLogger.Log(Object sender, LoggerEvent loggerEvent, String format, Object[] args)
2>...path to...\packages\MrAdvice.2.5.12\build\MrAdvice.targets(10,5): error :    at dnlib.DotNet.Pdb.PdbWriter.GetExternalInstructionOffset(CurrentMethod& info, MethodDef method, Instruction instr)
2>...path to...\packages\MrAdvice.2.5.12\build\MrAdvice.targets(10,5): error :    at dnlib.DotNet.Pdb.PdbWriter.WriteAsyncMethod(CurrentMethod& info, PdbAsyncMethod asyncMethod)
2>...path to...\packages\MrAdvice.2.5.12\build\MrAdvice.targets(10,5): error :    at dnlib.DotNet.Pdb.PdbWriter.Write(MethodDef method)
2>...path to...\packages\MrAdvice.2.5.12\build\MrAdvice.targets(10,5): error :    at dnlib.DotNet.Pdb.PdbWriter.Write()
2>...path to...\packages\MrAdvice.2.5.12\build\MrAdvice.targets(10,5): error :    at dnlib.DotNet.Writer.ModuleWriterBase.WritePdbFile()
2>...path to...\packages\MrAdvice.2.5.12\build\MrAdvice.targets(10,5): error :    at dnlib.DotNet.Writer.ModuleWriter.WriteFile()
2>...path to...\packages\MrAdvice.2.5.12\build\MrAdvice.targets(10,5): error :    at dnlib.DotNet.Writer.ModuleWriter.WriteImpl()
2>...path to...\packages\MrAdvice.2.5.12\build\MrAdvice.targets(10,5): error :    at dnlib.DotNet.Writer.ModuleWriterBase.Write(Stream dest)
2>...path to...\packages\MrAdvice.2.5.12\build\MrAdvice.targets(10,5): error :    at dnlib.DotNet.Writer.ModuleWriterBase.Write(String fileName)
2>...path to...\packages\MrAdvice.2.5.12\build\MrAdvice.targets(10,5): error :    at dnlib.DotNet.ModuleDef.Write(String filename, ModuleWriterOptions options)
2>...path to...\packages\MrAdvice.2.5.12\build\MrAdvice.targets(10,5): error :    at StitcherBoy.Reflection.ModuleManager.Write(String assemblyOriginatorKeyFile)
2>...path to...\packages\MrAdvice.2.5.12\build\MrAdvice.targets(10,5): error :    at StitcherBoy.Weaving.Build.AssemblyStitcher.Process(StringDictionary parameters, Guid buildID, DateTime buildTime, String entryAssemblyPath)
2>...path to...\packages\MrAdvice.2.5.12\build\MrAdvice.targets(10,5): error :    at StitcherBoy.Weaving.StitcherProcessor.Process(StringDictionary parameters, Guid buildID, DateTime buildTime, String entryAsemblyPath)
2>...path to...\packages\MrAdvice.2.5.12\build\MrAdvice.targets(10,5): error :    at StitcherBoy.Weaving.StitcherProcessor.Process(StringDictionary parameters, Guid buildID, DateTime buildTime, String entryAsemblyPath)
2>...path to...\packages\MrAdvice.2.5.12\build\MrAdvice.targets(10,5): error :    at StitcherTask`1.Run(Boolean fromExe)

**Comment by picrap** [comment]
What is the difference between the two assemblies?

**Comment by hhblaze** [comment]
Hmm, they are completely different. But all of them were weaved under 2.5.1

**Comment by picrap** [comment]
I've filed an issue (https://github.com/0xd4d/dnlib/issues/134); until it's found and solved, your best chance is to get back t an old version.

**Comment by hhblaze** [comment]
Clear, will do

**Comment by hhblaze** [comment]
I can test with new 0xd4d error messages, if you prepare a new release

**Comment by picrap** [comment]
2.5.13! :wink: Let me know.

**Comment by hhblaze** [comment]
Ok, smth like this

Severity	Code	Description	Project	File	Line	Suppression State
Error		Unhandled exception: dnlib.DotNet.Writer.ModuleWriterException: Async method instruction has been removed but it's still being referenced by PDB info: BP Instruction: IL_0088: ldarg.0, BP Method: System.Void GpsCarControl.Server.CoreAsModule.RemoteInterfacesImplementer/<>c__DisplayClass30_0/<<Yadro_CompaniesAndUser_AddUser>b__0>d::MoveNext() (0x0600085E), Current Method: System.Void GpsCarControl.Server.CoreAsModule.RemoteInterfacesImplementer/<>c__DisplayClass30_0/<<Yadro_CompaniesAndUser_AddUser>b__0>d::MoveNext′() (0x06000000)
   at dnlib.DotNet.DummyLogger.Log(Object sender, LoggerEvent loggerEvent, String format, Object[] args)
   at dnlib.DotNet.Pdb.PdbWriter.GetExternalInstructionOffset(CurrentMethod& info, MethodDef method, Instruction instr)
   at dnlib.DotNet.Pdb.PdbWriter.WriteAsyncMethod(CurrentMethod& info, PdbAsyncMethod asyncMethod)
   at dnlib.DotNet.Pdb.PdbWriter.Write(MethodDef method)
   at dnlib.DotNet.Pdb.PdbWriter.Write()
   at dnlib.DotNet.Writer.ModuleWriterBase.WritePdbFile()
   at dnlib.DotNet.Writer.ModuleWriter.WriteFile()
   at dnlib.DotNet.Writer.ModuleWriter.WriteImpl()
   at dnlib.DotNet.Writer.ModuleWriterBase.Write(Stream dest)
   at dnlib.DotNet.Writer.ModuleWriterBase.Write(String fileName)
   at dnlib.DotNet.ModuleDef.Write(String filename, ModuleWriterOptions options)
   at StitcherBoy.Reflection.ModuleManager.Write(String assemblyOriginatorKeyFile)
   at StitcherBoy.Weaving.Build.AssemblyStitcher.Process(StringDictionary parameters, Guid buildID, DateTime buildTime, String entryAssemblyPath)
   at StitcherBoy.Weaving.StitcherProcessor.Process(StringDictionary parameters, Guid buildID, DateTime buildTime, String entryAsemblyPath)
   at StitcherBoy.Weaving.StitcherProcessor.Process(StringDictionary parameters, Guid buildID, DateTime buildTime, String entryAsemblyPath)
   at StitcherTask`1.Run(Boolean fromExe)	GpsCarMagic.Server

**Comment by hhblaze** [comment]
and inside of this function AddUser there is such construction:

    res = Task.Run(async () => await _M._Users.AddUser
                                (args[0].DeserializeMe<Yadro.CompaniesAndUsers.YadroUser>(),
                                (args.Count > 1) ? (args[1] == "1" ? true : false) : false)
                              ).Result;

**Comment by hhblaze** [comment]
when I remark it - weaving works

**Comment by hhblaze** [comment]
Here is complex problem this function (where Task.Run is located) can't be async, but must call in a sync manner AddUser function (that is async)

**Comment by picrap** [comment]
`var res = Task.Run(async () => await test()).Result;`

**Comment by 0xd4d** [comment]
I noticed that there's a new method in that log message `MoveNext'()`. What happens to the original `MoveNext()` method and its instructions? If you check dnlib's `PdbWriter::GetExternalInstructionOffset()`, it should take the fast path (first `if` should be `true`), but it doesn't happen in this case. Instead it does a slow search for the instruction in the async BP method `MoveNext()` and doesn't find the instruction. So if the instruction wasn't deleted, it seems that it was moved to some other method.

**Comment by picrap** [comment]
Don't bother! That's my garbage 😄 And yes, basically, Mr Advice moves bodies from one method to another.

**Comment by 0xd4d** [comment]
The `PdbAsyncMethod` instance is in: `body.PdbMethod.AsyncMethod`

**Comment by picrap** [comment]
I can"t reproduce the issue. I pasted your code, but it builds correctly. Was it with version 2.5.13?

**Comment by picrap** [comment]
It was with a previous version. However, since I can reproduce the issue, I will use this to solve it.

**Comment by picrap** [comment]
Fixed in 2.5.14!

**Comment by hhblaze** [comment]
So, still generates error with 2.5.15 on the same place:
```C#
 public string Yadro_CompaniesAndUser_AddUser(string callerAddIn, List<string> args)
        {
            try
            {
                if (callerAddIn != "GM_UserManagement")
                    return "-1";

                long res = 0;

                res = Task.Run(async () => await _M._Users.AddUser
                                (args[0].DeserializeMe<Yadro.CompaniesAndUsers.YadroUser>(),
                                (args.Count > 1) ? (args[1] == "1" ? true : false) : false)
                              ).Result;

                return res.ToString();
            }
            catch (Exception ex)
            {
                
            }

            return "-1";
        }
```

Here is an error dump:
```C#
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error : Error while weaving method 'System.Void ...PROGRAM....Server.CoreAsModule.RemoteInterfacesImplementer/<>c__DisplayClass30_0/<<Yadro_CompaniesAndUser_AddUser>b__0>d::MoveNext()': System.Runtime.Serialization.SerializationException: Type 'dnlib.DotNet.MethodDefMD' in Assembly 'MrAdvice.Weaver, Version=2.0.0.0, Culture=neutral, PublicKeyToken=c0e7e6eab6f293d8' is not marked as serializable.
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error : 
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error : Server stack trace: 
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error :    at System.Runtime.Serialization.FormatterServices.InternalGetSerializableMembers(RuntimeType type)
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error :    at System.Runtime.Serialization.FormatterServices.<>c__DisplayClass9_0.<GetSerializableMembers>b__0(MemberHolder _)
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error :    at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error :    at System.Runtime.Serialization.FormatterServices.GetSerializableMembers(Type type, StreamingContext context)
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error :    at System.Runtime.Serialization.Formatters.Binary.WriteObjectInfo.InitMemberInfo()
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error :    at System.Runtime.Serialization.Formatters.Binary.WriteObjectInfo.InitSerialize(Object obj, ISurrogateSelector surrogateSelector, StreamingContext context, SerObjectInfoInit serObjectInfoInit, IFormatterConverter converter, ObjectWriter objectWriter, SerializationBinder binder)
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error :    at System.Runtime.Serialization.Formatters.Binary.WriteObjectInfo.Serialize(Object obj, ISurrogateSelector surrogateSelector, StreamingContext context, SerObjectInfoInit serObjectInfoInit, IFormatterConverter converter, ObjectWriter objectWriter, SerializationBinder binder)
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error :    at System.Runtime.Serialization.Formatters.Binary.ObjectWriter.Serialize(Object graph, Header[] inHeaders, __BinaryWriter serWriter, Boolean fCheck)
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error :    at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Serialize(Stream serializationStream, Object graph, Header[] headers, Boolean fCheck)
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error :    at System.Runtime.Remoting.Channels.CrossAppDomainSerializer.SerializeMessageParts(ArrayList argsToSerialize)
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error :    at System.Runtime.Remoting.Messaging.SmuggledMethodCallMessage..ctor(IMethodCallMessage mcm)
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error :    at System.Runtime.Remoting.Messaging.SmuggledMethodCallMessage.SmuggleIfPossible(IMessage msg)
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error :    at System.Runtime.Remoting.Channels.CrossAppDomainSink.SyncProcessMessage(IMessage reqMsg)
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error : 
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error : Exception rethrown at [0]: 
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error :    at System.Runtime.Remoting.Proxies.RealProxy.HandleReturnMessage(IMessage reqMsg, IMessage retMsg)
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error :    at System.Runtime.Remoting.Proxies.RealProxy.PrivateInvoke(MessageData& msgData, Int32 type)
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error :    at StitcherBoy.Logging.ILogging.WriteDebug(String format, Object[] parameters)
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error :    at ArxOne.MrAdvice.Weaver.AspectWeaver.WeaveAdvices(MarkedNode markedMethod, WeavingContext context)
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error :    at ArxOne.MrAdvice.Weaver.AspectWeaver.WeaveMethod(ModuleDef moduleDefinition, MarkedNode markedMethod, WeavingContext context)
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): warning : Advising async void method 'System.Void GpsCarMagic.Server.Form1::button1_Click(System.Object,System.EventArgs)' could confuse async advices. Consider switching its return type to async Task.
2>  MrAdvice 2.5.15 weaved module '...PROGRAM....Server, Version=1.385.2017.808, Culture=neutral, PublicKeyToken=null' (targeting framework .NET 4.5) in 1126ms
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error : Unhandled exception: dnlib.DotNet.Writer.ModuleWriterException: Async method instruction has been removed but it's still being referenced by PDB info: BP Instruction: IL_009C: ldarg.0, BP Method: System.Void ...PROGRAM....Server.CoreAsModule.RemoteInterfacesImplementer/<>c__DisplayClass30_0/<<Yadro_CompaniesAndUser_AddUser>b__0>d::MoveNext() (0x06000863), Current Method: System.Void ...PROGRAM....Server.CoreAsModule.RemoteInterfacesImplementer/<>c__DisplayClass30_0/<<Yadro_CompaniesAndUser_AddUser>b__0>d::MoveNext′() (0x06000000)
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error :    at dnlib.DotNet.DummyLogger.Log(Object sender, LoggerEvent loggerEvent, String format, Object[] args)
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error :    at dnlib.DotNet.Pdb.PdbWriter.GetExternalInstructionOffset(CurrentMethod& info, MethodDef method, Instruction instr)
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error :    at dnlib.DotNet.Pdb.PdbWriter.WriteAsyncMethod(CurrentMethod& info, PdbAsyncMethod asyncMethod)
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error :    at dnlib.DotNet.Pdb.PdbWriter.Write(MethodDef method)
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error :    at dnlib.DotNet.Pdb.PdbWriter.Write()
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error :    at dnlib.DotNet.Writer.ModuleWriterBase.WritePdbFile()
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error :    at dnlib.DotNet.Writer.ModuleWriter.WriteFile()
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error :    at dnlib.DotNet.Writer.ModuleWriter.WriteImpl()
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error :    at dnlib.DotNet.Writer.ModuleWriterBase.Write(Stream dest)
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error :    at dnlib.DotNet.Writer.ModuleWriterBase.Write(String fileName)
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error :    at dnlib.DotNet.ModuleDef.Write(String filename, ModuleWriterOptions options)
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error :    at StitcherBoy.Reflection.ModuleManager.Write(String assemblyOriginatorKeyFile)
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error :    at StitcherBoy.Weaving.Build.AssemblyStitcher.Process(StringDictionary parameters, Guid buildID, DateTime buildTime, String entryAssemblyPath)
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error :    at StitcherBoy.Weaving.StitcherProcessor.Process(StringDictionary parameters, Guid buildID, DateTime buildTime, String entryAsemblyPath)
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error :    at StitcherBoy.Weaving.StitcherProcessor.Process(StringDictionary parameters, Guid buildID, DateTime buildTime, String entryAsemblyPath)
2>...path...\packages\MrAdvice.2.5.15\build\MrAdvice.targets(10,5): error :    at StitcherTask`1.Run(Boolean fromExe)
========== Build: 1 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========
```

if this Task.Run(async () => await _M._Users.AddUse is remarked - compilation is OK

**Comment by picrap** [comment]
I'm sorry, this is (again) a logging problem, which I'm going to fix once for all in a dependency project. Until there, it should be fixed in 2.5.16.

**Comment by hhblaze** [comment]
Works for now, thank you!

**Comment by hhblaze** [comment]
I think the journey not finished yet, it compiles, but this function that contains
Task.Run(async () => await _M._Users.AddUser
fails to be run in RELEASE MODE

here is the ILDump
```C#

public string Yadro_CompaniesAndUser_AddUser(string callerAddIn, List<string> args)
{
	return (string)⚡Invocation.ProceedAspect(this, new object[]
	{
		callerAddIn,
		args
	}, methodof(RemoteInterfacesImplementer.Yadro_CompaniesAndUser_AddUser(string, List<string>)).MethodHandle, methodof(RemoteInterfacesImplementer.Yadro_CompaniesAndUser_AddUser′(string, List<string>)).MethodHandle, methodof(RemoteInterfacesImplementer.Yadro_CompaniesAndUser_AddUser″(object, object[])).MethodHandle);
}
```

```C#
[ExecutionPoint]
private string Yadro_CompaniesAndUser_AddUser′(string callerAddIn, List<string> args)
{
	RemoteInterfacesImplementer.<>c__DisplayClass30_0 <>c__DisplayClass30_ = new RemoteInterfacesImplementer.<>c__DisplayClass30_0();
	<>c__DisplayClass30_.args = args;
	try
	{
		string result;
		if (callerAddIn != "GM_UserManagement")
		{
			result = "-1";
			return result;
		}
		result = Task.Run<long>(delegate
		{
			RemoteInterfacesImplementer.<>c__DisplayClass30_0.<<Yadro_CompaniesAndUser_AddUser>b__0>d <<Yadro_CompaniesAndUser_AddUser>b__0>d;
			<<Yadro_CompaniesAndUser_AddUser>b__0>d.<>4__this = <>c__DisplayClass30_;
			<<Yadro_CompaniesAndUser_AddUser>b__0>d.<>t__builder = AsyncTaskMethodBuilder<long>.Create();
			<<Yadro_CompaniesAndUser_AddUser>b__0>d.<>1__state = -1;
			AsyncTaskMethodBuilder<long> <>t__builder = <<Yadro_CompaniesAndUser_AddUser>b__0>d.<>t__builder;
			<>t__builder.Start<RemoteInterfacesImplementer.<>c__DisplayClass30_0.<<Yadro_CompaniesAndUser_AddUser>b__0>d>(ref <<Yadro_CompaniesAndUser_AddUser>b__0>d);
			return <<Yadro_CompaniesAndUser_AddUser>b__0>d.<>t__builder.Task;
		}).Result.ToString();
		return result;
	}
	catch (Exception ex)
	{
		_Log.AddCoreException("GpsCarControl.Server.CoreAsModule.RemoteInterfacesImplementer", "Yadro_CompaniesAndUser_AddUser", ex, "");
	}
	return "-1";
}
```

```C#
private static object Yadro_CompaniesAndUser_AddUser″(object obj, object[] array)
{
	return ((RemoteInterfacesImplementer)obj).Yadro_CompaniesAndUser_AddUser′((string)array[0], (List<string>)array[1]);
}
```

**Comment by picrap** [comment]
What fails, exactly?

**Comment by hhblaze** [comment]
Since its Release mode and no debug is available, currently I see that async function is executed, but result is never returned back... I will think further how to make it for you more convenient

**Comment by hhblaze** [comment]
I could emulate it, grab the attached project.
Calling async task with the parameters make it hanging

var task = System.Threading.Tasks.Task.Run(async () => await AddWhatever(p1,p2));

[TestAsync (2).zip](https://github.com/ArxOne/MrAdvice/files/1225058/TestAsync.2.zip)

**Comment by snekbaev** [comment]
@hhblaze does that attached project compile with the latest v2.5.19 for you? I'm getting:

```
Severity	Code	Description	Project	File	Line	Suppression State
Error		Unhandled exception: dnlib.DotNet.Writer.ModuleWriterException: Instruction was removed from the body but is referenced from PdbScope: IL_0040: ldarg.0
   at dnlib.DotNet.DummyLogger.Log(Object sender, LoggerEvent loggerEvent, String format, Object[] args)
   at dnlib.DotNet.Pdb.WindowsPdb.WindowsPdbWriter.CurrentMethod.GetOffset(Instruction instr)
   at dnlib.DotNet.Pdb.WindowsPdb.WindowsPdbWriter.WriteAsyncMethod(CurrentMethod& info, PdbAsyncMethodCustomDebugInfo asyncMethod)
   at dnlib.DotNet.Pdb.WindowsPdb.WindowsPdbWriter.Write(MethodDef method, List`1 cdiBuilder)
   at dnlib.DotNet.Pdb.WindowsPdb.WindowsPdbWriter.Write()
   at dnlib.DotNet.Writer.ModuleWriterBase.WriteWindowsPdb(PdbState pdbState)
   at dnlib.DotNet.Writer.ModuleWriterBase.WritePdbFile()
   at dnlib.DotNet.Writer.ModuleWriter.WriteFile()
   at dnlib.DotNet.Writer.ModuleWriter.WriteImpl()
   at dnlib.DotNet.Writer.ModuleWriterBase.Write(Stream dest)
   at dnlib.DotNet.Writer.ModuleWriterBase.Write(String fileName)
   at dnlib.DotNet.ModuleDef.Write(String filename, ModuleWriterOptions options)
   at StitcherBoy.Reflection.ModuleManager.Write(String assemblyOriginatorKeyFile)
   at StitcherBoy.Weaving.Build.AssemblyStitcher.Process(StringDictionary parameters, Guid buildID, DateTime buildTime, String entryAssemblyPath)
   at StitcherBoy.Weaving.StitcherProcessor.Process(StringDictionary parameters, Guid buildID, DateTime buildTime, String entryAsemblyPath)
   at StitcherBoy.Weaving.StitcherProcessor.Process(StringDictionary parameters, Guid buildID, DateTime buildTime, String entryAsemblyPath)
   at StitcherTask`1.Run(Boolean fromExe)	TestAsync			
```

**Comment by snekbaev** [comment]
I commented out everything in the `Test` class except the `StartTask` method which I changed into
```
public void StartTask()
            {
var r = Task.Run(
                                 async () =>
                                 {
                                     await Task.Delay( 1000 );
                                     return 100;
                                    
                                 } ).Result;
 }
```
it seems that it fails to compile if there is an `async` lambda

**Comment by picrap** [comment]
I've fixed a problem with `Tasks` in 2.8.8. Can someone tell me if the issue above is solved?

**Comment by hhblaze** [comment]
Works in the latest version!
![image](https://user-images.githubusercontent.com/486781/66266544-c0a5bf80-e826-11e9-8644-8146a89b7773.png)
